tutorials/arvo/cli-tutorial.md

```diff
@@ -0,0 +1,420 @@
++++
+title = "Command line interface apps"
+weight = 13
+template = "doc.html"
++++
+
+# Command line interface apps
+
+## Introduction
+
+In this walkthrough we will go in-depth on how to build command line interface (CLI)
+applications in Urbit using the `shoe` library.
+
+There are three CLI apps that currently ship with urbit - `%dojo`, `%chat-cli`,
+and `%shoe`. You should be familiar with the former two, the latter is an example
+app that shows off how the `shoe` library works that we will be looking at
+closely. These are all Gall apps, and as such their source can be found in the `app/` folder of your `%home` desk.
+
+In [the `shoe` library](#the-shoe-library) we take a closer look at the `shoe` library and its
+cores and how they are utilized in CLI apps. Then in [the `sole`
+library](#the-sole-library) we look at what `shoe` effects ultimately break down
+into. Finally in [`%shoe` app walkthrough](#shoe-app-walkthrough) we explore
+the functionality of the `%shoe` app and then go through the code line-by-line.
+
+This tutorial can be
+considered to be an application equivalent of the [Hoon school
+lesson](@/docs/tutorials/hoon/generators.md#ask) on `sole` and `%ask`
+generators, which only covers the bare minimum necessary to write generators
+that take user input.
+
+
+## The `shoe` library {#the-shoe-library}
+
+Here we describe how sessions are identified, the specialized `card`s that Gall agents
+with the `shoe` library are able to utilize, and the different cores of `/lib/shoe.hoon` and their purpose.
+
+### Session identifiers
+
+An app using the `shoe` library will automatically track sessions by their `sole-id=@ta`.
+These are opaque identifiers generated by the connecting client. An app using the `shoe`
+library may be connected to by a local or remote ship in order to send commands,
+and each of these connections is assigned a unique `@ta` that identifies the
+ship and which session on that ship if there are multiple.
+
+### `%shoe` `card`s
+
+Gall agents with the `shoe` library are able to utilize `%shoe` `card`s. These
+additions to the standard set of `cards` have the following shape.
```

Comment:
This is super nitpicky, and it's more of a comment than an actual "pls change this" suggestion, but:

```suggestion
additions to the standard set of `cards` have the following shape:
```

I always think `:` is more appropriate, whenever some list or example follows, and you're actively pointing to it. This applies to other such cases in this document, too.

Reply:
I no good grammar, but I do recall that my habit in this regard was formed by never letting a sentence end without a period being highest priority, and that doing this sometimes resulted in that depending on the structure of the list that followed. I'll look up what the proper style is supposed to be.